src/Orleans/Streams/Internal/StreamDirectory.cs

```diff
@@ -20,7 +20,13 @@ internal StreamDirectory()
 
         internal IAsyncStream<T> GetOrAddStream<T>(StreamId streamId, Func<IAsyncStream<T>> streamCreator)
         {
-            return allStreams.GetOrAdd(streamId, _ => streamCreator()) as IAsyncStream<T>;
+            var stream = allStreams.GetOrAdd(streamId, _ => streamCreator()) as IAsyncStream<T>;
+            if (stream == null)
+            {
+                throw new Runtime.OrleansException($"Stream type mismatch. Please use GetStream<T>(Guid id, string streamNamespace) with the correct type, which is the type application first referenced the stream with.");
```

Comment:
Fixed. and added tests.
Sample exception is like:

Message: Orleans.Runtime.OrleansException : Stream type mismatch. A stream can only support a single type of data. The generic type of the stream requested does not match the previously requested type Orleans.Streams.StreamImpl`1[System.Int32].